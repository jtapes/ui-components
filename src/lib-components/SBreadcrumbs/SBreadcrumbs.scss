@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-breadcrumbs {
  display: inline-flex;
  align-items: center;
  overflow: hidden;
  width: 100%;
  text-overflow: ellipsis;

  &__item {
    white-space: nowrap;
  }

  &__separator {
    display: inline-flex;
    align-items: center;
    margin: 0 $size--12;
  }
}
