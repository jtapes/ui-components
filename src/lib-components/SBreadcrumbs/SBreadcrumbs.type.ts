export type BreadcrumbType = {
  label: string
  link?: string
}
