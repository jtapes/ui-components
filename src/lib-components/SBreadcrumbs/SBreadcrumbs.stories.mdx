import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SBreadcrumbs from '@/lib-components/SBreadcrumbs/SBreadcrumbs.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Breadcrumbs"
  component={SBreadcrumbs}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    breadcrumbs: {
      control: 'array',
      defaultValue: [
        { label: 'Каталог', link: 'catalog' },
        { label: 'Сортовой прокат', link: 'sortovoy-prokat' },
        { label: 'Лист горячетканый горячетканый горячетканый' },
      ]
    },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=9%3A284',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Хлебные крошки

<Description of={SBreadcrumbs} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SBreadcrumbs },
  template: `<s-breadcrumbs :breadcrumbs="breadcrumbs" />`,
})

<Canvas withToolbar withSource="none">
  <Story name="SBreadcrumbs">
    {Template.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
<s-breadcrumbs :breadcrumbs="[{ label: 'Каталог', link: 'catalog' }, { label: 'Сортовой прокат', link: 'sotrovoy-prokat' }, { label: 'Лист горячетканый горячетканый горячетканый' }]">
  <template #separator>
    <span>-></span>
  </template>
</s-breadcrumbs>
  `}
/>

<ArgsTable of={SBreadcrumbs} />
