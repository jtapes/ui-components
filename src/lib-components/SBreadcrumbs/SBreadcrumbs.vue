<template>
  <div class="s-breadcrumbs text-sm color-font-secondary">
    <template v-for="(crumb, index) in breadcrumbs">
      <div :key="crumb.label" class="s-breadcrumbs__item">
        <template v-if="crumb.link">
          <!-- @slot Слот для вставки компонентов формирования ссылок -->
          <slot
            name="link"
            :crumb="crumb"
            :index="index"
            :crumbFormat="crumbFormat"
          >
            <a
              :href="crumb.link"
              class="link color-font-secondary"
              @click.prevent="pageChange(crumb.link)"
            >
              {{ crumbFormat(crumb.label, index) }}
            </a>
          </slot>
        </template>
        <span v-else>
          {{ crumbFormat(crumb.label, index) }}
        </span>
      </div>
      <div
        v-if="!isLastCrumb(index)"
        :key="index"
        class="s-breadcrumbs__separator"
      >
        <!-- @slot Слот для разделителя между крошками -->
        <slot name="separator">
          <svg-icon name="arrow-16" class="icon-16" />
        </slot>
      </div>
    </template>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import { BreadcrumbType } from './SBreadcrumbs.type'

/** Компонент для отображения хлебных крошек */
@Component({
  components: {
    SvgIcon,
  }
})
export default class SBreadcrumbs extends Vue {
  /** Массив с хлебными крошками */
  @Prop({
    type: Array,
    validator(value: BreadcrumbType[]): boolean {
      return value.every((breadcrumb) =>
        Object.prototype.hasOwnProperty.call(breadcrumb, 'label')
      )
    },
  })
  breadcrumbs!: BreadcrumbType[]

  isLastCrumb(index: number): boolean {
    return index + 1 === this.breadcrumbs.length
  }

  crumbFormat(crumb: string, index: number): string {
    const maxLengthOfCrumb = 40
    if (this.isLastCrumb(index) && crumb.length > maxLengthOfCrumb) {
      return `${crumb.slice(0, maxLengthOfCrumb)}...`
    } else {
      return crumb
    }
  }

  pageChange(link: string): void {
    this.$emit('page-change', link)
  }
}
</script>
