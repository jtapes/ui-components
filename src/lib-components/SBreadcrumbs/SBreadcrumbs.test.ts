import { mount } from '@vue/test-utils'
import SBreadcrumbs from '@/lib-components/SBreadcrumbs/SBreadcrumbs.vue'

describe('SBreadcrumbs', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SBreadcrumbs, {
      propsData: {
        breadcrumbs: [
          { label: 'Каталог', link: 'catalog' },
          { label: 'Сортовой прокат', link: 'sortovoy-prokat' },
          { label: 'Лист горячетканый горячетканый горячетканый' },
        ],
      },
      stubs: ['svg-icon', 'nuxt-link'],
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
