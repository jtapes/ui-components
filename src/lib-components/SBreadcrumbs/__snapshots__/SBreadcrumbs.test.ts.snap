// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SBreadcrumbs Is Vue snapshot 1`] = `
<div
  class="s-breadcrumbs text-sm color-font-secondary"
>
  <div
    class="s-breadcrumbs__item"
  >
    <a
      class="link color-font-secondary"
      href="catalog"
    >
      
            Каталог
          
    </a>
  </div>
   
  <div
    class="s-breadcrumbs__separator"
  >
    <svg-icon-stub
      class="icon-16"
      name="arrow-16"
    />
  </div>
  <div
    class="s-breadcrumbs__item"
  >
    <a
      class="link color-font-secondary"
      href="sortovoy-prokat"
    >
      
            Сортовой прокат
          
    </a>
  </div>
   
  <div
    class="s-breadcrumbs__separator"
  >
    <svg-icon-stub
      class="icon-16"
      name="arrow-16"
    />
  </div>
  <div
    class="s-breadcrumbs__item"
  >
    <span>
      
        Лист горячетканый горячетканый горячетка...
      
    </span>
  </div>
   
  <!---->
</div>
`;
